plugins/scaffolder-react/src/next/extensions/createScaffolderFormDecorator.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { AnyApiRef } from '@backstage/core-plugin-api';
+import { JsonValue } from '@backstage/types';
+import { z } from 'zod';
+
+/** @alpha */
+export type ScaffolderFormDecoratorContext<TInput> = {
+  input: TInput;
+  formState: Record<string, JsonValue>;
+
+  setFormState: (
+    fn: (currentState: Record<string, JsonValue>) => Record<string, JsonValue>,
+  ) => void;
+  setSecrets: (
+    fn: (currentState: Record<string, string>) => Record<string, string>,
+  ) => void;
+};
+
+/** @alpha */
+export type ScaffolderFormDecorator<
+  TInputSchema extends { [key in string]: (zImpl: typeof z) => z.ZodType } = {},
+  TDeps extends { [key in string]: AnyApiRef } = { [key in string]: AnyApiRef },
+  TInput extends {} = {
+    [key in keyof TInputSchema]: z.infer<ReturnType<TInputSchema[key]>>;
```

Comment:
Aim to have the defaults for these be the "any" version, so the first one would default to `{ [key in string]: (zImpl: typeof z) => z.ZodType }`. With that it should be possible to get rid of the `AnyScaffolderFormDecorator` and just use `ScaffolderFormDecorator` without type args instead